packages/expo/src/DataMigrationHelper.ts

```diff
@@ -0,0 +1,87 @@
+import Constants from 'expo-constants';
+import * as FileSystem from 'expo-file-system';
+import { Platform } from 'react-native';
+
+export interface ConflictResolver {
+  (legacyFile: string, currentFile: string): Promise<void>;
+}
+
+export const LOCK_FILE_NAME = 'migrationLock64537438658125';
+
+export function getLegacyDocumentDirectoryAndroid(): string | null {
+  if (Platform.OS !== 'android' || FileSystem.documentDirectory == null) {
+    return null;
+  }
+  // expo-file-system decodes paths so we need to encode twice
+  let id: string = encodeURIComponent(encodeURIComponent(Constants.manifest.id));
+  const oldFilesDirectory = `${FileSystem.documentDirectory}ExperienceData/${id}/`;
+  return oldFilesDirectory;
+}
+
+export const noopResolve: ConflictResolver = async (legacyFile: string, currentFile: string): Promise<void> => {
+  // do nothing! leave legacy and current file
+};
+
+async function treeSearch(relativePath: string, legacyPath: string, newPath: string, resolveConflict: ConflictResolver): Promise<void> {
+  const currentNewPath = `${newPath}${relativePath}`;
+  const currentLegacyPath: string = `${legacyPath}${relativePath}`;
+  const legacyPathInfo = await FileSystem.getInfoAsync(currentLegacyPath);
+  const newPathInfo = await FileSystem.getInfoAsync(currentNewPath);
+
+  if (legacyPathInfo.exists && !newPathInfo.exists) {
+    await FileSystem.copyAsync({
+      from: currentLegacyPath,
+      to: currentNewPath,
+    });
+    await FileSystem.deleteAsync(currentLegacyPath);
+    return;
+  }
+
+  if (legacyPathInfo.isDirectory) {
+    const children = await FileSystem.readDirectoryAsync(currentLegacyPath);
+    for (let child of children) {
+      await treeSearch(relativePath + `${child}/`, legacyPath, newPath, resolveConflict);
+    }
+  } else {
+    await resolveConflict(currentLegacyPath, currentNewPath);
+  }
+}
+
+async function doesOldFilesDirectoryContainLock(path: string): Promise<boolean> {
+  const children = await FileSystem.readDirectoryAsync(path);
+  return children.indexOf(LOCK_FILE_NAME) > -1;
+}
+
+async function addLockToOldFilesDirectory(path: string): Promise<void> {
+  await FileSystem.writeAsStringAsync(path + LOCK_FILE_NAME, 'lock');
+}
+
+export async function migrateFilesFromLegacyDirectoryAsync(resolveConflict?: ConflictResolver): Promise<void> {
+  const { appOwnership } = Constants;
+  if (Platform.OS !== 'android' || appOwnership !== 'standalone') {
+    return;
+  }
+  const oldFilesDirectory = getLegacyDocumentDirectoryAndroid();
+  const newFilesDirectory = FileSystem.documentDirectory;
+  if (oldFilesDirectory == null || newFilesDirectory == null) {
+    return;
+  }
+
+  const oldFilesDirectoryInfo = await FileSystem.getInfoAsync(<string>oldFilesDirectory);
+  const doesOldFilesDirectoryExist = oldFilesDirectoryInfo['exists'];
+
+  if (!doesOldFilesDirectoryExist || await doesOldFilesDirectoryContainLock(oldFilesDirectory)) {
+    return;
+  }
+
+  if (resolveConflict == null) {
```

Comment:
according to TS type, `resolveConflict` can be `undefined` but not `null`. This should probably just check `if (!resolveConflict)` instead.

Reply:
In JS it is still possible to pass null as a `resolveConflict`.  
```js
console.log(null == undefined)
```
output : true

Reply:
I would follow @esamelson idea 😉 The user is responsible to ensure that the passed argument is a valid one according to function signature. @Szymon20000 following your idea, we should also ensure that passed argument comforts `ConflictResolver` interface and is not not, for example of type `number` 😉
Nowadays comparison via `==` is strongly discouraged and every linter/other tool would warn you about it.

Reply:
I just thought that someone can pass null as an argument. Passing a number is highly unlikely. However, I see I was wrong.